src/opera/compare/template_comparer.py

```diff
@@ -0,0 +1,137 @@
+        new_context = TemplateContext(node1, node2,
+                                      context.workdir1, context.workdir2)
+        return self.compare_item(node1, node2, comparisons, new_context)
+
+    def compare_requirement(self, req1, req2, context):
+        comparisons = {
+            "target": Comparison(self.compare_target),
+            "relationship": Comparison(self.compare_relationship)
+        }
+        return self.compare_item(req1, req2, comparisons, context)
+
+    def compare_capability(self, cabability1, cabability2, context):
+        comparisons = {
+            "properties": MapComparison(self.compare_capability_property),
+            "attributes": MapComparison(self.compare_none)
+        }
+        return self.compare_item(cabability1, cabability2, comparisons,
+                                 context)
+
+    def compare_relationship(self, rel1, rel2, context):
+        comparisons = {
+            "types": Comparison(self.compare_types),
+            "properties": MapComparison(self.compare_value),
+            "attributes": MapComparison(self.compare_none),
+            "interfaces": MapComparison(self.compare_interface)
+        }
+        new_context = TemplateContext(rel1, rel2,
+                                      context.workdir1, context.workdir2)
+        return self.compare_item(rel1, rel2,
+                                 comparisons,
+                                 new_context)
+
+    def compare_interface(self, interface1, interface2, context):
+        comparisons = {
+            "operations": MapComparison(self.compare_operation)
+        }
+        return self.compare_item(interface1, interface2, comparisons, context)
+
+    def compare_operation(self, operation1, operation2, context):
+        comparisons = {
+            "primary": Comparison(self.compare_file),
+            "inputs": MapComparison(self.compare_value),
+            "dependencies": ListComparison(self.compare_file,
+                                           self.get_filename),
+            "artifacts": ListComparison(self.compare_file,
+                                        self.get_filename)
+        }
+        return self.compare_item(operation1, operation2, comparisons, context)
+
+    def compare_value(self, val1, val2, context):
+        try:
+            eval1 = val1.eval(context.template1, "")
+        except DataError:
+            eval1 = None
+        try:
+            eval2 = val2.eval(context.template2, "")
+        except DataError:
+            eval2 = None
+        return eval1 == eval2, [eval1, eval2]
+
+    def compare_types(self, types1, types2, context):
+        type1 = types1[0]
+        type2 = types2[0]
+        return type1 == type2, [type1, type2]
+
+    def compare_capability_property(self, prop1, prop2, context):
+        return prop1.data == prop2.data, \
+               [prop1.data, prop2.data]
+
+    def compare_none(self, attr1, attr2, context):
+        # do not compare attribute values
```

Comment:
Why are attributes skipped?

Reply:
Attributes are usually assigned values during deployment (e.g. IP addresses). If we compare lets say identical templates attributes would certainly differ, while technically the deployments would be the identical.

Reply:
This makes sense then and can be kept as it is.